bfd/BFD_Enhancement_HLD.md

```diff
@@ -0,0 +1,453 @@
+    sonic(config-bfd-peer)# transmit-interval 200
+```                                              
+**Show Output**                                  
+Below output shows BFD session in UP state established for BGP protocol.
+
+```
+    sonic# show bfd peer
+    BFD Peers:          
+        peer 10.0.0.103 interface Ethernet204
+            ID: 2                            
+            Remote ID: 5                     
+            Status: up                       
+            Uptime: 30 second(s)             
+            Diagnostics: ok                  
+            Remote diagnostics: ok           
+            Local timers:                    
+                Receive interval: 300ms      
+                Transmission interval: 300ms 
+                Echo transmission interval: disabled
+            Remote timers:
+                Receive interval: 200ms
+                Transmission interval: 200ms
+                Echo transmission interval: 0ms
+```
+**BFD Counters**
+Below output shows BFD counter for particular BFD session
+```
+sonic# show bfd peer 192.168.0.1 counters
+     peer 192.168.0.1
+             Control packet input: 126 packets
+             Control packet output: 247 packets
+             Echo packet input: 2409 packets
+             Echo packet output: 2410 packets
+             Session up events: 1
+             Session down events: 0
+             Zebra notifications: 4
+```
+# 6 Warm Boot Support
+Planned/unplanned warm-boot of a BGP container can be achieved by running BGP in GR mode.
+
+When BGP container is rebooted, remote BGP neighbour enters GR helper mode when BFD indicate timeout and GR helper mode is enabled. If the flag in BFD control packet indicate that the BFD in remote neighbour is not control plane independent, BFD session down event can be a trigger to BGP to enter helper mode. In GR helper mode remote BGP neighbor should not delete the routes learnt through BGP neighbour and keep the forwarding plane intact until GR timeout.
+After warm-boot is completed BGP will re-establish all the sessions and trigger BFD to establish corresponding BFD sessions.
+
+# 7 BFD packet trapping to CPU
+
+ 
+BFD packets are classified based on packet pattern defined in RFC 5880. Trapping of Single-Hop and Multi-hop BFD packets for IPv4 and IPv6 address family is supported.
+
+  
+
+BFD related control plane QoS can be configured by user via 00-copp.config.json file. User can specify which CPU queue a protocol (e.g. BFD) is trapped to and policing parameters to rate limit the protocol packet. Below is an example of the BFD protocol trapping COPP (Control Plane Policing) configuration in 00-copp.config.json file:
+
+```
+{
+"COPP_TABLE:trap.group.bfd": {
+"trap_ids": "bfd,bfdv6",
+"trap_action":"trap",
+"trap_priority":"5",
+"queue": "5",
+"meter_type":"packets",
+"mode":"sr_tcm",
+"cir":"60000",
+"cbs":"60000",
+"red_action":"drop"
+},
+"OP": "SET"
+}
+```
+
+# 7 Scalability
+ - No of sessions: 64
```

Comment:
Is 64 a hard limit? Do you have any test result for the maximum number of sessions?

Reply:
No 64 it's not a hard limit. We have tested with 64 session but i do not have log to attach.